[WPE] Measure browser binary size as part of the benchmark tests on the RPi bots

clopez opened the pull request:
#### 1dfcb0e870a6cb8cf0a3ec2790d72394242af13d
<pre>
[WPE] Measure browser binary size as part of the benchmark tests on the RPi bots
<a href="https://bugs.webkit.org/show_bug.cgi?id=294625">https://bugs.webkit.org/show_bug.cgi?id=294625</a>

Reviewed by Nikolas Zimmermann.

This patch implements a new benchmark plan named &apos;browser-binary-size&apos; that can
be run with the script browserperfdash-benchmark and which will gather the size
in bytes of the browser and the relevant browser libraries and send those to
the dashboard. By relevant browser libraries we understand those that are not
provided by the system or the third-party libraries system (jhbuild/flatpak)

An example of the data that it sends in the case of WPE MiniBrowser is this
one:

Size:Total: 179MB
     MiniBrowser:Size: 3.51MB
     libWPEPlatform-2.0.so.0:Size: 3.62MB
     libWPEWebKit-2.0.so.1:Size: 172MB

Note: it sends the data in bytes, above is the pretty-print version of it.

The idea is to run this &apos;browser-binary-size&apos; report continously as part of
the battery of benchmark tests that we run on the build.webkit.org CI and
send this data to the browserperfdash dashboard to keep a track of this
sizes and be able to easily tell when the sizes changed.

In order to accomodate this new type of benchmark plan in the design of
browserperfdash a new type of plan plugins is created. Those plugins
consist on independent python files that register themselves with a
plugin-name string and a entry-point function that receive as parameters
the browser object (from run-benchmark) and the path to a temporal file
where it should write the data with the results in json format.

For now only one of this plugins is there: &apos;browser-binary-size&apos;, but
the design allows to easily ship more random plugin-plans in the future.

This new type of plan-plugins are meant to measure something in the browser
(like binary-size) but they don&apos;t need to execute the browser itself.
This contrasts with the existing run-benchmark plans, which are meant to
execute the browser on a specific webpage that runs some benchmarks within
the browser.

Also, a few related or needed features are implemented, like:

  - Support on browserperfdash-benchmark for saving the json results sent to
the server in a local directory
  - Support on browserperfdash-benchmark for reading and dumping to stdout the
parsed information from those saved json results

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_results.py:
(BenchmarkResults):
* Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
(parse_args):
(BrowserPerfDashRunner.__init__):
(BrowserPerfDashRunner._get_plan_version_hash):
(BrowserPerfDashRunner):
(BrowserPerfDashRunner._load_and_list_plan_plugins):
(BrowserPerfDashRunner._is_benchmark_runner_plan):
(BrowserPerfDashRunner.is_browserperfdash_runner_plan):
(BrowserPerfDashRunner.available_plans):
(BrowserPerfDashRunner._run_benchmark_runner_plan):
(BrowserPerfDashRunner._run_plan):
(BrowserPerfDashRunner._save_to_results_directory):
(BrowserPerfDashRunner.run):
(read_results_json):
(main):
(BrowserPerfDashRunner._get_test_version_string): Deleted.
* Tools/Scripts/webkitpy/browserperfdash/plans/browser_binary_size.py: Added.
(generate_json_for_benchmark):
(get_basenames_and_sizes):
(get_browser_relevant_objects_glib):
(run):
(register):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.get_browser_path):
(GtkPort.run_minibrowser):
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort.get_browser_path):
(WPEPort.run_minibrowser):
* Tools/Scripts/webkitpy/port/wpe_unittest.py:
(WPEPortTest.test_browser_cog_parameters_platform_override_via_environ):
(WPEPortTest):
(WPEPortTest.test_get_browser_path):

Canonical link: <a href="https://commits.webkit.org/296425@main">https://commits.webkit.org/296425@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/944106da42ecd8b278e8a64b4a44d0a11a863c2f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28167 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18591 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/113715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110469 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36719 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/113715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111454 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/97734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/113715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/107938 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/22311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/15872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/58430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/15925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/116836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35560 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26205 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/116836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35933 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/94008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/116836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/36125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/13889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17519 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35462 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35171 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38523 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36852 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu commented on Tools/Scripts/browserperfdash-benchmark:
```diff
@@ -29,6 +29,16 @@ import sys
 
 
+if sys.platform.startswith('linux'):
+    import os
```
I don't think we need to defer import for builtin modules like `os`

clopez replied:
sure, thanks for the suggestion.
I moved this change to https://github.com/WebKit/WebKit/pull/46957 after splitting up this PR as requested, and applied there the suggested change

dewei-zhu commented on Tools/Scripts/webkitpy/binary_bundling/ldd.py:
```diff
@@ -65,10 +96,11 @@ def _get_libs_and_interpreter(self, object):
                 line = line.split('=>')[1].strip()
                 if 'not found' in line:
-                    raise RuntimeError('ldd can not resolve all dependencies for object %s.' % object)
+                    raise RuntimeError('ldd can not resolve all dependencies for object "%s". Error in line "%s".' % (object, orig_line))
```
Nit: probably want to convert to f-string since we are touching this code

clopez replied:
sure, thanks for the suggestion.
I moved this change to #46955 after splitting up this PR as requested, and applied there the suggested change

dewei-zhu commented on Tools/Scripts/webkitpy/port/wpe.py:
```diff
@@ -147,7 +150,7 @@ def run_minibrowser(self, args):
 
         if self.browser_name() == "cog":
-            miniBrowser = self.cog_path_to('launcher', 'cog')
+            miniBrowser = self.get_browser_path("cog")
```
nit: any reason to prefer double-quote over single quote? It looks like most of this path uses single-quote.

clopez replied:
no reason other than consistency with the surrounding. In this function `run_minibrowser(self, args)` all strings are quoted with double quotes, the two ones I replaced were the only ones using single quotes out of 13 inside this function.

nikolaszimmermann commented on Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
```diff
@@ -111,6 +124,13 @@ def __init__(self, args):
             _log.info('Will send the benchmark data as if it was generated on date: {date}'.format(date=date_str))
             self._result_data['timestamp'] = timestamp
 
+        # Gethe browser version if needed
```
Typo gethe

nikolaszimmermann commented on Tools/Scripts/webkitpy/browserperfdash/plans/browser_binary_size.py:
```diff
@@ -0,0 +1,86 @@
+from webkitpy.port import configuration_options, platform_options, factory
+from webkitpy.binary_bundling.ldd import SharedObjectResolver
+
+PLUGIN_NAME = "browser-binary-size"
```
Neat idea with the plugins!

nikolaszimmermann reviewed (approved):
Great job @clopez.

clopez committed:
[WPE] Measure browser binary size as part of the benchmark tests on the RPi bots
https://bugs.webkit.org/show_bug.cgi?id=294625

Reviewed by Nikolas Zimmermann.

This patch implements a new benchmark plan named 'browser-binary-size' that can
be run with the script browserperfdash-benchmark and which will gather the size
in bytes of the browser and the relevant browser libraries and send those to
the dashboard. By relevant browser libraries we understand those that are not
provided by the system or the third-party libraries system (jhbuild/flatpak)

An example of the data that it sends in the case of WPE MiniBrowser is this
one:

Size:Total: 179MB
     MiniBrowser:Size: 3.51MB
     libWPEPlatform-2.0.so.0:Size: 3.62MB
     libWPEWebKit-2.0.so.1:Size: 172MB

Note: it sends the data in bytes, above is the pretty-print version of it.

The idea is to run this 'browser-binary-size' report continously as part of
the battery of benchmark tests that we run on the build.webkit.org CI and
send this data to the browserperfdash dashboard to keep a track of this
sizes and be able to easily tell when the sizes changed.

In order to accomodate this new type of benchmark plan in the design of
browserperfdash a new type of plan plugins is created. Those plugins
consist on independent python files that register themselves with a
plugin-name string and a entry-point function that receive as parameters
the browser object (from run-benchmark) and the path to a temporal file
where it should write the data with the results in json format.

For now only one of this plugins is there: 'browser-binary-size', but
the design allows to easily ship more random plugin-plans in the future.

This new type of plan-plugins are meant to measure something in the browser
(like binary-size) but they don't need to execute the browser itself.
This contrasts with the existing run-benchmark plans, which are meant to
execute the browser on a specific webpage that runs some benchmarks within
the browser.

Also, a few related or needed features are implemented, like:

  - Support on browserperfdash-benchmark for saving the json results sent to
the server in a local directory
  - Support on browserperfdash-benchmark for reading and dumping to stdout the
parsed information from those saved json results

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_results.py:
(BenchmarkResults):
* Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
(parse_args):
(BrowserPerfDashRunner.__init__):
(BrowserPerfDashRunner._get_plan_version_hash):
(BrowserPerfDashRunner):
(BrowserPerfDashRunner._load_and_list_plan_plugins):
(BrowserPerfDashRunner._is_benchmark_runner_plan):
(BrowserPerfDashRunner.is_browserperfdash_runner_plan):
(BrowserPerfDashRunner.available_plans):
(BrowserPerfDashRunner._run_benchmark_runner_plan):
(BrowserPerfDashRunner._run_plan):
(BrowserPerfDashRunner._save_to_results_directory):
(BrowserPerfDashRunner.run):
(read_results_json):
(main):
(BrowserPerfDashRunner._get_test_version_string): Deleted.
* Tools/Scripts/webkitpy/browserperfdash/plans/browser_binary_size.py: Added.
(generate_json_for_benchmark):
(get_basenames_and_sizes):
(get_browser_relevant_objects_glib):
(run):
(register):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.get_browser_path):
(GtkPort.run_minibrowser):
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort.get_browser_path):
(WPEPort.run_minibrowser):
* Tools/Scripts/webkitpy/port/wpe_unittest.py:
(WPEPortTest.test_browser_cog_parameters_platform_override_via_environ):
(WPEPortTest):
(WPEPortTest.test_get_browser_path):

Canonical link: https://commits.webkit.org/296425@main
